src/csharp/Microsoft.Spark/Utils/AssemblyLoader.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.Spark.Utils
+{
+    internal static class AssemblyLoader
+    {
+        internal static Func<string, Assembly> LoadFromFile { get; set; } = Assembly.LoadFrom;
+
+        internal static Func<string, Assembly> LoadFromName { get; set; } = Assembly.Load;
+
+        private static readonly Dictionary<string, Assembly> s_assemblyCache =
+            new Dictionary<string, Assembly>();
+
+        private static readonly string[] s_searchPaths =
+            new[] { Directory.GetCurrentDirectory(), AppDomain.CurrentDomain.BaseDirectory };
+
+        private static readonly string[] s_extensions =
+            RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ?
+                    new[] { ".dll", ".exe", ".ni.dll", ".ni.exe" } :
+                    new[] { ".dll", ".ni.dll" };
+
+        private static readonly object s_cacheLock = new object();
+
+        /// <summary>
+        /// Return the cached assembly, otherwise attempt to load and cache the assembly
+        /// in the following order:
+        /// 1) Load the assembly from disk using assemblyFileName.
+        /// 2) Search for assemblyName in the current app domain.
```

Comment:
question: what's the reason for this ordering?

Reply:
1) if we already have a hint as to what the filename should be, then we will use this value first.
2) We will check if this assemblyname is currently can be found within the current domain/context.  Otherwise we'll try the other extensions to look for the assembly.

Reply:
Looks like the hint (the filename) is always there. If I ask it differently, what happens if we try to do 2) first?

Reply:
To make the check stricter, we can just error when the assemblyFileName is not present and not worry about condition 2.

Reply:
Updated and removed 2)